### PR TITLE
Make logging to synchronous for daemons

### PR DESCRIPTION
**Pull request by sakshamgarg**
#### Bug/feature Description
* *Making logging back to synchronous*

#### Affected Platform(s)
* *All*

#### Cause / Analysis / Design
* *If one of the daemons die and doesn't drop a core file, with asynchronous logging, there is a possibility of loosing last few log messages.*

#### Solution Description
* *Adding "fflush()" function to make logging to synchronous*
* *Pulling out changes as done in the PR: https://github.com/PBSPro/pbspro/pull/943*

#### Testing logs/output
* *Since we are reverting back to old state, there is no need for any specific testing. Appveyor and Travis should be fine.*

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/lib/Liblog/pbs_log.c [hunk 1]
```diff
@@ -931,6 +931,7 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
 			     eventtype & ~PBSEVENT_FORCE, msg_daemonname,
 			     class_names[objclass], objname, text);
 
+		(void)fflush(logfile);
```
Comment (mike0042) [inline src/lib/Liblog/pbs_log.c]: This could have a significant performance impact in some cases. Would it not be better to make this configurable?

Comment (subhasisb) [reply src/lib/Liblog/pbs_log.c]: HI @mike0042 -  the fflush existed in earlier code which we removed recently, without realizing that the called used were the buffered libc calls that would keep a log message in the application buffers (without the fflush) and if the application crashed, one might then lose critical/crucial log information that might point out what the application was doing during its last moments of life. This seriously hampers functionality. The gain in performance by removing the fflush is only seen in very slow filesystems like a home on NFS etc. but customers always have the option of going to syslog in those cases. Besides the performance gain (far we measured) was barely modest. Hence, we want to restore functionality.

Comment (mike0042) [reply src/lib/Liblog/pbs_log.c]: Hello @subhasisb - If we have numbers to backup the performance impact is minimal in nearly all normal cases of operation, I don't see the point of making it configurable. I withdraw my request, and thank you for the explanation.

Comment (sakshamgarg) [reply src/lib/Liblog/pbs_log.c]: hi @mike0042  I will post the performance numbers in the PR by today

**Comment by sakshamgarg** [comment]
Hi @mike0042 , I tested the performance in normal case on a test machine.
The job run rate after making logging to asynchronous is 185 jobs/sec. While after making logging to synchronous, the run rate is 175 jobs/sec. There is not much difference.

**Comment by mike0042** [comment]
Thank you @sakshamgarg. I have marked the discussion as resolved and approved your PR.

**Comment by sakshamgarg** [comment]
Thanks @mike0042 and @nithinj for your approval. I have rebased. Kindly take this further.

**Pull request merged**